contracts/misc/UniswapV3OracleWrapper.sol

```diff
@@ -141,4 +174,221 @@ contract UniswapV3OracleWrapper is IAtomicPriceAggregator {
+            );
+        } else if (oracleData.tokenBDecimal > oracleData.tokenADecimal) {
+            // multiple by 10^(decimalB - decimalA) to preserve price in wei
+            oracleData.sqrtPriceX96 = uint160(
+                (SqrtLib.sqrt(
+                    (oracleData.tokenAPrice *
+                        (10 **
+                            (18 +
+                                oracleData.tokenBDecimal -
+                                oracleData.tokenADecimal))) /
+                        (oracleData.tokenBPrice)
+                ) * 2**96) / 10E9
+            );
+        } else {
+            // multiple by 10^(decimalA - decimalB) to preserve price in wei then divid by the same number
+            oracleData.sqrtPriceX96 = uint160(
+                (SqrtLib.sqrt(
+                    (oracleData.tokenAPrice *
+                        (10 **
+                            (18 +
+                                oracleData.tokenADecimal -
+                                oracleData.tokenBDecimal))) /
+                        (oracleData.tokenBPrice)
+                ) * 2**96) /
+                    10 **
+                        (9 +
+                            oracleData.tokenADecimal -
+                            oracleData.tokenBDecimal)
+            );
+        }
+
+        return oracleData;
+    }
+
+    function _getPositionData(uint256 tokenId)
+        internal
+        view
+        returns (UinswapV3PositionData memory)
+    {
+        UinswapV3PositionData memory positionData;
+        (
+            ,
+            ,
+            address token0,
+            address token1,
+            uint24 fee,
+            int24 tickLower,
+            int24 tickUpper,
+            uint128 liquidity,
+            uint256 positionFeeGrowthInside0LastX128,
+            uint256 positionFeeGrowthInside1LastX128,
+            uint256 tokensOwed0,
+            uint256 tokensOwed1
+        ) = UNISWAP_V3_POSITION_MANAGER.positions(tokenId);
+
+        return
+            UinswapV3PositionData({
+                tokenA: token0,
+                tokenB: token1,
+                fee: fee,
+                tickLower: tickLower,
+                tickUpper: tickUpper,
+                liquidity: liquidity,
+                positionFeeGrowthInsideALastX128: positionFeeGrowthInside0LastX128,
+                positionFeeGrowthInsideBLastX128: positionFeeGrowthInside1LastX128,
+                tokensOwedA: tokensOwed0,
+                tokensOwedB: tokensOwed1
+            });
+    }
+
+    function _getLiquidityAmount(
+        uint160 sqrtRatioX96,
+        UinswapV3PositionData memory positionData
+    ) internal pure returns (uint256 token0Amount, uint256 token1Amount) {
+        (token0Amount, token1Amount) = LiquidityAmounts.getAmountsForLiquidity(
+            sqrtRatioX96,
+            TickMath.getSqrtRatioAtTick(positionData.tickLower),
+            TickMath.getSqrtRatioAtTick(positionData.tickUpper),
+            positionData.liquidity
```

Comment:
this is the virtual liquidity which cannot reflect users' position, I think we should use `SqrtPriceMath.getAmount0Delta` instead